crates/ty_ide/src/rename.rs

```diff
@@ -0,0 +1,773 @@
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::tests::{CursorTest, IntoDiagnostic, cursor_test};
+    use insta::assert_snapshot;
+    use ruff_db::diagnostic::{Annotation, Diagnostic, DiagnosticId, LintName, Severity, Span};
+    use ruff_db::files::FileRange;
+    use ruff_text_size::Ranged;
+
+    impl CursorTest {
+        fn prepare_rename(&self) -> String {
+            let Some(range) = can_rename(&self.db, self.cursor.file, self.cursor.offset) else {
+                return "Cannot rename".to_string();
+            };
+
+            format!("Can rename symbol at range {range:?}")
+        }
+
+        fn rename(&self, new_name: &str) -> String {
+            let Some(rename_results) =
+                rename(&self.db, self.cursor.file, self.cursor.offset, new_name)
+            else {
+                return "Cannot rename".to_string();
+            };
+
+            if rename_results.is_empty() {
+                return "No locations to rename".to_string();
+            }
+
+            self.render_diagnostics(rename_results.into_iter().enumerate().map(
+                |(i, ref_item)| -> RenameResult {
+                    RenameResult {
+                        index: i,
+                        file_range: FileRange::new(ref_item.file(), ref_item.range()),
+                    }
+                },
+            ))
+        }
+    }
+
+    struct RenameResult {
+        index: usize,
+        file_range: FileRange,
+    }
+
+    impl IntoDiagnostic for RenameResult {
+        fn into_diagnostic(self) -> Diagnostic {
+            let mut main = Diagnostic::new(
+                DiagnosticId::Lint(LintName::of("rename")),
+                Severity::Info,
+                format!("Rename location {}", self.index + 1),
+            );
+            main.annotate(Annotation::primary(
+                Span::from(self.file_range.file()).with_range(self.file_range.range()),
+            ));
+
+            main
+        }
+    }
+
+    #[test]
+    fn test_prepare_rename_parameter() {
+        let test = cursor_test(
+            "
+def func(<CURSOR>value: int) -> int:
+    value *= 2
+    return value
+
+value = 0
+",
+        );
+
+        assert_snapshot!(test.prepare_rename(), @"Can rename symbol at range 10..15");
+    }
+
+    #[test]
+    fn test_rename_parameter() {
+        let test = cursor_test(
+            "
+def func(<CURSOR>value: int) -> int:
+    value *= 2
+    return value
+
+func(value=42)
+",
+        );
+
+        assert_snapshot!(test.rename("number"), @r"
+        info[rename]: Rename location 1
+         --> main.py:2:10
+          |
+        2 | def func(value: int) -> int:
+          |          ^^^^^
+        3 |     value *= 2
+        4 |     return value
+          |
+
+        info[rename]: Rename location 2
+         --> main.py:3:5
+          |
+        2 | def func(value: int) -> int:
+        3 |     value *= 2
+          |     ^^^^^
+        4 |     return value
+          |
```

Comment:
Same as what I mentioned on another PR. You could add multiple annotations to the same diagnostic instead where the first annotation highlights the identifier that the user wants to rename and all other annotations show where the renames are necessary.

That would reduce the redundant information in tests